app/scripts/modules/core/src/api/mock/expectedRequest.ts

```diff
@@ -0,0 +1,145 @@
+import { ReceivedRequest } from 'core/api/mock/receivedRequest';
+import { deferred, isSuccessStatus, UrlArg, Verb } from './mockHttpUtils';
+import { SETTINGS } from 'core/config';
+import { isEqual, isMatch } from 'lodash';
+
+function parseParams(queryString: string): Record<string, string | string[]> {
+  const paramTuples = queryString.split('&').map((param) => param.split('=')) as Array<[string, string]>;
+
+  return paramTuples.reduce((paramsObj, [key, value]) => {
+    const currentValue = paramsObj[key];
+    if (typeof currentValue === 'string') {
+      paramsObj[key] = [currentValue, value];
+    } else if (Array.isArray(currentValue)) {
+      paramsObj[key] = paramsObj[key].concat(value);
+    } else {
+      paramsObj[key] = value;
+    }
+    return paramsObj;
+  }, {} as Record<string, string | string[]>);
+}
+
+/**
+ * This internal class represents an HTTP request that is expected to be made by the code under test.
+ * Instances are created by MockHttpClient.expect -- this class should not be used directly.
+ *
+ * This class exposes a promise (this.fulfilledDeferred.promise).
+ * The promise will be resolved when the code under test makes the expected HTTP request.
+ *
+ * ExpectedRequest also stores the (optional) mock response status/data to return to the code under test.
+ */
+export class ExpectedRequest implements IExpectBuilder {
```

Comment:
I would have expected this class to also support a notion of a non-response from the server. Does Deck make the assumption that the backend is always accessible and returning responses of some kind (as opposed to just timing out)? Do we ever assert the behavior of Deck when the backend services are simply unavailable?

Reply:
... I suppose a timeout in browser land is exposed as a 408? Maybe there's nothing to do on this use case.

Reply:
that's a great question.  deck doesn't handle timeouts at this layer and thus we don't provide testing capability.  There is an interceptor in the AngularJS $http client that will retry timed out requests.  If retries are exhausted, the request is rejected (no http codes involved).  When we move from $http to something else (fetch?) we will likely re-create this capability and write tests at that layer.